File: plugins/togetherai-skills/skills/together-sandboxes/scripts/execute_with_session.ts
#!/usr/bin/env -S npx tsx
/**
 * Together AI Code Interpreter — Execute Code with Session Reuse (TypeScript SDK)
 *
 * Run Python code in a sandboxed environment, reuse sessions to persist state,
 * upload files, and handle display outputs.
 *
 * Usage:
 *     npx tsx execute_with_session.ts
 *
 * Requires:
 *     npm install together-ai
 *     export TOGETHER_API_KEY=your_key
 */

import Together from "together-ai";

const client = new Together();

async function executeCode(
  code: string,
  sessionId?: string,
  files?: { name: string; encoding: string; content: string }[]
): Promise<{ sessionId: string | null; outputs: any[] }> {
  const params: any = { code, language: "python" };
  if (sessionId) params.session_id = sessionId;
  if (files) params.files = files;

  const response = await client.codeInterpreter.execute(params);

  if (response.errors) {
    console.error(`Errors: ${JSON.stringify(response.errors)}`);
    return { sessionId: null, outputs: [] };
  }

  const outputs: any[] = [];
  for (const output of response.data.outputs) {
    if (output.type === "stdout" || output.type === "stderr") {
      console.log(`  [${output.type}] ${output.data}`);
    } else if (output.type === "error") {
      console.log(`  [error] ${output.data}`);
    } else if (output.type === "display_data" || output.type === "execute_result") {
      const keys =
        typeof output.data === "object" ? Object.keys(output.data) : [];
      console.log(`  [${output.type}] ${JSON.stringify(keys)}`);
    }
    outputs.push({ type: output.type, data: output.data });
  }

  return { sessionId: response.data.session_id, outputs };
}

async function listSessions() {
  const response = await client.codeInterpreter.sessions.list();
  for (const session of response.data?.sessions ?? []) {
    console.log(
      `  Session ${session.id}: ${session.execute_count} executions, expires ${session.expires_at}`
    );
  }
}

async function main() {
  // --- Example 1: Single execution ---
  console.log("=== Single execution ===");
  const result = await executeCode(
    'print("Hello from Together Code Interpreter!")'
  );
  const sessionId = result.sessionId!;
  console.log(`Session ID: ${sessionId}\n`);

  // --- Example 2: Reuse session (state persists) ---
  console.log("=== Session reuse -- define variable ===");
  await executeCode("x = 42\nprint(f'Set x = {x}')", sessionId);

  console.log("\n=== Session reuse -- access variable ===");
  await executeCode("print(f'x is still {x}')", sessionId);

  // --- Example 3: Data analysis ---
  console.log("\n=== Data analysis ===");
  await executeCode(
    `
import numpy as np

data = np.random.randn(1000)
print(f"Mean: {data.mean():.4f}")
print(f"Std:  {data.std():.4f}")
print(f"Min:  {data.min():.4f}")
print(f"Max:  {data.max():.4f}")
`,
    sessionId
  );

  // --- Example 4: Chart generation ---
  console.log("\n=== Chart generation ===");
  await executeCode(
    `
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import numpy as np

x = np.linspace(0, 10, 100)
plt.figure(figsize=(8, 4))
plt.plot(x, np.sin(x), label='sin(x)')
plt.plot(x, np.cos(x), label='cos(x)')
plt.legend()
plt.title('Trig Functions')
plt.show()
`,
    sessionId
  );

  // --- Example 5: File upload ---
  console.log("\n=== File upload ===");
  await executeCode("!python myscript.py", undefined, [
    {
      name: "myscript.py",
      encoding: "string",
      content: "import sys\nprint(f'Hello from {sys.argv[0]}!')",
    },
  ]);

  // --- List active sessions ---
  console.log("\n=== Active sessions ===");
  await listSessions();
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: plugins/togetherai-skills/skills/together-sandboxes/scripts/execute_with_session.py
#!/usr/bin/env python3
"""
Together AI Code Interpreter — Execute Code with Session Reuse (v2 SDK)

Run Python code in a sandboxed environment, reuse sessions to persist state,
and handle file outputs.

Usage:
    python execute_with_session.py

Requires:
    uv pip install "together>=2.0.0"
    export TOGETHER_API_KEY=your_key
"""

from together import Together

client = Together()


def execute_code(code: str, session_id: str | None = None) -> dict:
    """Execute Python code, optionally in an existing session."""
    response = client.code_interpreter.execute(
        code=code,
        language="python",
        **({"session_id": session_id} if session_id else {}),
    )

    if response.errors:
        print(f"Errors: {response.errors}")
        return {"session_id": None, "outputs": [], "errors": response.errors}

    outputs = []
    for output in response.data.outputs:
        if output.type in ("stdout", "stderr"):
            print(f"  [{output.type}] {output.data}")
            outputs.append({"type": output.type, "data": output.data})
        elif output.type == "error":
            print(f"  [error] {output.data}")
            outputs.append({"type": "error", "data": output.data})
        elif output.type in ("display_data", "execute_result"):
            print(f"  [{output.type}] {list(output.data.keys()) if isinstance(output.data, dict) else output.data}")
            outputs.append({"type": output.type, "data": output.data})

    return {
        "session_id": response.data.session_id,
        "outputs": outputs,
        "errors": None,
    }


def list_sessions():
    """List active code interpreter sessions."""
    response = client.code_interpreter.sessions.list()
    for s in response.data.sessions:
        print(f"  Session {s.id}: {s.execute_count} executions, expires {s.expires_at}")
    return response.data.sessions


if __name__ == "__main__":
    # --- Example 1: Single execution ---
    print("=== Single execution ===")
    result = execute_code("print('Hello from Together Code Interpreter!')")
    session_id = result["session_id"]
    print(f"Session ID: {session_id}\n")

    # --- Example 2: Reuse session (state persists) ---
    print("=== Session reuse — define variable ===")
    execute_code("x = 42\nprint(f'Set x = {x}')", session_id=session_id)

    print("\n=== Session reuse — access variable ===")
    execute_code("print(f'x is still {x}')", session_id=session_id)

    # --- Example 3: Data analysis with packages ---
    print("\n=== Data analysis ===")
    execute_code(
        """
import numpy as np

data = np.random.randn(1000)
print(f"Mean: {data.mean():.4f}")
print(f"Std:  {data.std():.4f}")
print(f"Min:  {data.min():.4f}")
print(f"Max:  {data.max():.4f}")
""",
        session_id=session_id,
    )

    # --- Example 4: Generate a chart and retrieve as base64 PNG ---
    # plt.show() with the Agg backend does not reliably produce display_data
    # outputs. Instead, save the figure to a buffer and encode it explicitly.
    print("\n=== Chart generation ===")
    chart_result = execute_code(
        """
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import numpy as np
import base64, io

x = np.linspace(0, 10, 100)
fig, ax = plt.subplots(figsize=(8, 4))
ax.plot(x, np.sin(x), label='sin(x)')
ax.plot(x, np.cos(x), label='cos(x)')
ax.legend()
ax.set_title('Trig Functions')

buf = io.BytesIO()
fig.savefig(buf, format='png', dpi=150)
buf.seek(0)
png_b64 = base64.b64encode(buf.read()).decode()
plt.close(fig)
print(f"chart_png_base64:{png_b64}")
""",
        session_id=session_id,
    )

    # Extract and save the chart locally
    import base64 as _b64

    for out in chart_result["outputs"]:
        if out["type"] == "stdout" and "chart_png_base64:" in out["data"]:
            b64_str = out["data"].split("chart_png_base64:", 1)[1].strip()
            with open("trig_chart.png", "wb") as f:
                f.write(_b64.b64decode(b64_str))
            print(f"  Chart saved to trig_chart.png ({len(b64_str)} bytes b64)")

    # --- List active sessions ---
    print("\n=== Active sessions ===")
    list_sessions()
